feat: Add support for Monolog 3.x

jeromegamez opened the pull request:
Hey everyone 👋,

this is a follow-up to #2302 and aims to add support for all currently available monolog/monolog versions.

* `AppEngineFlexHandlerV3` and `AppEngineFlexFormatterV3` are the new implementations for Monolog 3.x.
* The `formatPayload` method in the `FormatterTrait` should support arrays and the new `LogRecord` class

#### Prerequisites for this PR to be mergeable

* ~https://github.com/phpDocumentor/Reflection/pull/247  is merged but not released (this PR uses ^5.x-dev for now)~
* ~https://github.com/googleapis/gax-php/pull/389 is not reviewed/merged/released (this PR uses my PR branch from over there for now)~
* ~https://github.com/GoogleCloudPlatform/grpc-gcp-php/pull/45 is not (yet) merged - if it's just because of the CLA, I can create a new PR there~

#### Things to look into

* The ReflectionHandlerV5 class is a 90% copy of the ReflectionHandlerV4 - perhaps this can be deduplicated

#### Notes

* There are deprecation warnings that I was not yet able to tackle yet:
  * `opis/closure`
    > Deprecated: Opis\Closure\SerializableClosure implements the Serializable interface, which is deprecated. Implement __serialize() and __unserialize() instead (or in addition, if support for old PHP versions is necessary) in /Users/jg/Code/opensource/google-cloud-php/vendor/opis/closure/src/SerializableClosure.php on line 18

:octocat:

jeromegamez commented:
I'll ignore the conventionalcommits.org tests for now, but it's good to see that the "Generate Documentation" job fails, this shows that https://github.com/googleapis/google-cloud-php/blob/main/Core/src/Testing/Reflection/ReflectionHandlerFactory.php needs updating to support phpdocumentor/reflection ^5

jeromegamez commented:
The reflection handlers should probably be refactored to avoid much-duplicated code, I will get to that once I'm back at my desk in a few hours

jeromegamez commented:
Alright, this is a rabbit hole - when I fix one problem, another one pops up, that's why I'm gradually working my way through the classes and force pushing all the time.

The good news is that we'll probably have PHP 8.1 compatibility when we're through (if that's even possible)

jeromegamez commented:
Update: Sorry for all the force pushes - I needed to make a bunch of preparations in order to be able to run the tests without throwing errors with PHP 8.1 before adding Monolog 3 to the mix.

There's currently a blocker with a `ReadOnly` class because it collides with the reserved `readonly` in PHP 8.1.

This has already been addressed and a fix is on its way with https://github.com/protocolbuffers/protobuf/pull/10041

Until then I think this PR is on hold, but I'm watching the repo and will continue as soon as it's through... perhaps google/gax und phpdocumentor/reflection will have released new versions as well 🤞

bshaffer commented on .github/workflows/release.yml:
```diff
@@ -19,6 +19,8 @@ jobs:
             php-version: '7.4'
             ini-values: memory_limit=2048M
+          env:
+            COMPOSER_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
Please remove these, why would you need to modify our secrets to add monolog support? This is a huge security concern...

jeromegamez replied:
This wasn't meant for optimization but only as a temporary change in order to be able to pull in my google/gax fork. Without it, I ran into permission errors, for example here: https://github.com/googleapis/google-cloud-php/runs/6853003115?check_suite_focus=true

My plan was to revert these changes once the fork isn't needed anymore 😅

alexander-schranz replied:
Can confirm it doesn't hurt to have this in your github actions. It just make it easier for contributors as they can test against forks. i keep it in every project to make things easier.

jeromegamez replied:
In the meantime, the tokens have been removed 😅

jeromegamez commented:
Whew! Alright, 20 action runs later, I think I'm getting somewhere - the test suites finally don't exit with errors anymore 🥳

Due to the two PRs not being merged/released yet, this one is not mergeable as well, but if someone wants to try this PR in their projects, that would be the best test 😅. And of course, nobody has reviewed the changes yet 😬.

(The Docs Generator Job is still failing, and I can't figure out why. If you have an idea, please let me know, otherwise I'll give it a closer look tomorrow with a rested brain 💪)

bshaffer commented on composer.json:
```diff
@@ -59,11 +59,11 @@
         "squizlabs/php_codesniffer": "3.*",
-        "phpdocumentor/reflection": "^3.0||^4.0",
-        "symfony/console": "^3.0",
+        "phpdocumentor/reflection": "^3.0||^4.0||^5.x-dev",
```
Why are we adding support for `5.x-dev`? Is this because `4.x` of `phpdocumentor/reflection` requires `psr/log:1.0`? 

This is a monorepo, and is discouraged from being installed. So we should be able to support Monolog 3.x without having to add support for a major version (and especially a major version in `-dev`). We would do this by adding support in the `Core` library, and relying on the individual package tests (see https://github.com/googleapis/google-cloud-php/pull/5325).

bshaffer reviewed (commented):
Hello! Wow, thanks for your work here. It looks like a whole lot of work. I am not sure why the docs generation is failing either, but I imagine it has to do with adding support for the new major version of `phpdocumentor/reflection:5.x-dev`. This is something we should be able to avoid doing (as mentioned in my comment below). 

We will be removing this portion of our docs generation shortly, so adding support for a new major dev version is not something we want to do (and it should be avoidable). Unfortunately we'll need to wait for the "individual package test" to get merged (https://github.com/googleapis/google-cloud-php/pull/5325) first, but that should allow us to test the changes you've made here to Core.

jeromegamez commented:
I had to add the dev branch of the reflection library because its dependency of `psr/log` 1 prevented Monolog from installing, since it needs version 3. I hoped my PR over there would have been released, but removing that part of the doc generation altogether sounds even better! 

I will rebase this PR with main to see if everything still works 🤞🏻.

Do you see a chance to merge and release https://github.com/GoogleCloudPlatform/grpc-gcp-php/pull/45 in the near future? This is another blocker 😅

jeromegamez commented:
Alright! I managed to green up all checks 🥳 

Concerning the jump to a new major version: `phpdocumentor/reflection` is exclusively used as a dev-dependency in the following places (I think you mentioned that already, @bshaffer):

* `Core` (only tests)
* docs generation

Although it's unfortunate to have to rely on a not-yet-released dev dependency, we already hat one dev dependency before (which is now removed):

<img width="508" alt="CleanShot 2022-10-11 at 00 39 38@2x" src="https://user-images.githubusercontent.com/67554/194962677-9bc80b09-3415-4b03-a084-26cff8d6a385.png">

One out, one in? 😅 

#5325 could still be applied afterward, and I have a notification on the next `phpdocumentor/reflection` stable 5.x release.

To re-iterate the `COMPOSER_TOKEN: ${{ secrets.GITHUB_TOKEN }}` topic: this doesn't expose a (new) secret. The token is available on every workflow run, and its only purpose is to avoid running into rate limits when fetching composer dependencies.

jeromegamez commented:
I also want to add that many changes seemingly are not related to Monolog, but without them, the tests failed on my machine with PHP 8.1.

jamiehollern commented:
Great work @jeromegamez. I'm relying on this PR to make one of the APIs work with Symfony 6, so many thanks for your efforts. Is there anything you would like support with?

bshaffer commented on Logging/src/LogMessageProcessorFactory.php:
```diff
@@ -0,0 +1,47 @@
+    /**
+     * @return LogMessageProcessor
+     */
+    public function getLogMessageProcessor()
```
Let's rename this to `build` and make this static to follow the same convention we use in [HttpHandlerFactory](https://github.com/googleapis/google-auth-library-php/blob/main/src/HttpHandler/HttpHandlerFactory.php):
```suggestion
    public static function build()
```

jeromegamez replied:
Done!

bshaffer commented on Logging/tests/Unit/LoggerInterfaceTest.php:
```diff
@@ -0,0 +1,135 @@
+use Psr\Log\LoggerInterface;
+use Psr\Log\LogLevel;
+
+abstract class LoggerInterfaceTest extends TestCase
```
This makes me nervous... it looks like this class was copied over from the PSR tests?

All the methods you've copied over are `public`, so why wouldn't we just implemented whatever changes we need in `PsrLoggerCompatibilityTest.php` instead, since that's the only test that extends this `LoggerInterfaceTest`? 

Also, what changes are required? I can't tell since you copied everything over.

jeromegamez replied:
I copied them over because `psr/log` 1.1 had a TestLogger which has been removed in 2.0 and 3.0.

This way, all `psr/log` versions are testable.

However, in the meantime, [fig/log-test](https://github.com/php-fig/log-test#versions) has been started, but support for `psr/log` 2.0/3.0 has not yet been released. 

I'm going to require that package with `^1.0|^2.0-dev` in the hope both can be used and remove the `-dev` once it's released (this should happen within the week).

jeromegamez replied:
I think so, too, at least for the moment. I'll do it later tonight! 🤞🏻

jeromegamez replied:
I think I'll have to wait until fig/log-test 1.1 is released - or it's too late for me, and I can't figure out how to add the requirement to use the released version or the dev-version, depending on which `psr/log` version is installed. This, at least, doesn't work for me right now:

```
"fig/log-test": "1.0.*||dev-main#1d0c40f",
```

jeromegamez replied:
The PR now uses the latest release of `fig/log-test` - since the current release only supports `psr/log ^1`, `monolog/monolog` is again fixed to 2.x until `fig/log-test` receives the planned release.

I've tested it locally with the main branch, and the tests fail because the method signatures of the `PsrLogger` class are incompatible with Monolog v3, so I have some more work to do until next week.

bshaffer commented on composer.json:
```diff
@@ -60,11 +60,11 @@
         "erusev/parsedown": "^1.6",
         "vierbergenlars/php-semver": "^3.0",
-        "symfony/lock": "3.3.x-dev#1ba6ac9",
+        "symfony/lock": "^3.3||^4.4||^5.4||^6.2",
```
It would be best to avoid any/all changes to `composer.json`. Yes, this means that monolog 3.x won't be available to the monorepo (e.g. `google/cloud`), but as long as it's available to `google/cloud-core`, then the individual packages will be installable, which is the important thing.

jeromegamez replied:
That's good news, but I think the docs generation GitHub Workflow will then break, because it uses the root `composer.json` 😕

But I hope I'm wrong about this 🤞🏻

jamiehollern replied:
@bshaffer Maybe I'm missing something, but without support for Monolog 3, `google/cloud` is not compatible with Symfony 6. I would like to use the Google Vision API in my Symfony 6 application but can't unless this is supported.

jeromegamez replied:
But I have to add... it makes local development on the root project as a whole more complicated when the dependencies on the root project are not updated. Do you work on each component separately in a sandboxed IDE?

When setting up the project for my PRs, I used https://github.com/googleapis/google-cloud-php/blob/main/CONTRIBUTING.md as the guideline…

jeromegamez replied:
Could you please confirm once more that I should leave the root `composer.json` untouched? At the moment it just™ widens the requirements, which (I think) shouldn't have negative effects.

bshaffer replied:
@jamiehollern see https://github.com/googleapis/google-cloud-php/issues/5273 the root package `google/cloud` has been discouraged from use for years, and soon will be turned down entirely. You can use the Google Vision API in your application by installing [`google/cloud-vision`](https://github.com/googleapis/google-cloud-php-vision).

We don't want to support all the additional dependencies implied by this PR, and our docs generation is quite brittle and may have adverse consequences updating to a new major (and dev) version of `phpdocumentor`. 

I am happy to look into a potentially better configuration of dependencies in the future, after we remove the documentation component which is coming soon. But for now, it would be best to leave it out, so we can get this PR merged sooner rather than later.

jeromegamez replied:
@bshaffer I'm a little bit in a deadlock here… 

* `google/cloud-logging` requires a released `google/cloud-core` version
* The currently released `google/cloud-core` supports `"monolog/monolog": "^1.0|^2.0"`, which prevents `google/cloud-logging` from installing Monolog V3
* Widening and releasing the requirement in `google/cloud` would break `google/cloud` for anyone using PHP 8.1 until `google/cloud-logging` is updated and released (and projects using `google/cloud-*` have updated all their dependencies

I understand the concern about not wanting to maintain the dependencies in the root `composer.json`, but I'm not sure how we can proceed here:

* The individual package tests will fail or at least not cover the Monolog V3 changes with tests
* If we don't update the root `composer.json`, we can't test and release both components simultaneously because the new features won't get picked up
* If we do update the root `composer.json`, there's the maintenance burden, but the individual package tests will still fail

A solution could be to move everything log-related from https://github.com/googleapis/google-cloud-php/tree/main/Core/src/Logger as well as the Monolog dependency into the Logging component and add Monolog V3 Support afterward, but I'm pretty sure this would break BC.

I'd appreciate your advice 🙏! If there's a chance to discuss this more directly/synchronously instead of in long-winded PR comment threads, you can find my contact details on my GH profile page!

jeromegamez replied:
Then again, our previous conversation on this topic has brought us even more comments 😂 

<img width="780" alt="CleanShot 2022-10-13 at 00 15 20@2x" src="https://user-images.githubusercontent.com/67554/195458198-7c1fbf46-cdcd-4c12-a943-32d959403572.png">

bshaffer replied:
I do not think we are understanding each other @jamiehollern. I am not incorrect, let me try to explain.

When you run `composer install google/cloud-vision`, the `composer.json` for this repository is not effected. Composer is looking at [this file](https://github.com/googleapis/google-cloud-php/blob/main/Vision/composer.json), which uses `google/cloud-core` as a dependency, and that is determined by [this file](https://github.com/googleapis/google-cloud-php/blob/main/Core/composer.json). 

So the conflict you are getting is absolutely as a result of this repository, and this PR here can potentially fix it. But the root composer.json ([this file](https://github.com/googleapis/google-cloud-php/blob/main/composer.json)) and the package `google/cloud` (which installs this entire repository) is not involved.

@jeromegamez I see your point. Thank you for explaining this. I think I know how to fix the issue, but it will need to be in another PR. Basically, we'll need to configure composer to install `google/cloud-core` from the `Core` directory instead of from packagist. This is something I've done before.

jeromegamez replied:
@bshaffer That's good news, thank you!

Another PR is even good, that gives me more time to clean up this one and move all the changes regarding better PHP 8.1 support and smaller fixes not related to Core and Logging to new PRs 💪🏻

bshaffer replied:
@jeromegamez try this https://github.com/googleapis/google-cloud-php/commit/2e0f7101e026bad55a618efca008e4516ebb9273

jeromegamez replied:
Thanks! I had to adapt the script a little bit (https://github.com/jeromegamez/google-cloud-php/blob/monolog-3/.github/run-package-tests.sh) because using the local path didn't work unless setting `"canonical": false` ([Composer Repository Priorities](https://getcomposer.org/doc/articles/repository-priorities.md#making-repositories-non-canonical)).

The package tests are now executed, but the Monolog Upgrade is not covered because the PHP Version is fixed to 8.0 - fixing it to 8.1 would render the previous Monolog Versions untested.

I think the way to go would be to include the package tests in the test matrix, but this could get expensive (I'm thinking time and Packagist rate limits 🤔 )

bshaffer replied:
I've updated it now to be a little smarter (it uses repo options to set a version, which gets around the problem you mentioned)

jeromegamez replied:
Thank you, this looks much better and the tests almost pass - the one error I see is left is https://github.com/googleapis/google-cloud-php/actions/runs/3246076562/jobs/5324410527#step:4:1480

```
1) Google\Cloud\Spanner\Tests\Snippet\Batch\BatchClientTest::testPubSubExample
Prophecy\Exception\Call\UnexpectedCallException: Unexpected method call on Double\ConnectionInterface\P7:
  - publishMessage(
        ["topic" => "...", "messages" => [["attributes" => ["snapshot" => xxx"]]]]
    )
expected calls were:
  - publishMessage(
        [..., exact("messages") => exact([["attributes" => ["snapshot" => "xxx"]]]), ...]
    )
  - pull(
        *
    )
```

I looked into it, but I'm not very versed in Prophecy, so I'm not (yet) sure how to fix it.

jeromegamez replied:
@bshaffer I pulled in your latest changes from #5542 and reverted all changes in the root `composer.json` file. Thank you for your work on this; I couldn't have done it myself 😅 🙏

jeromegamez commented:
Good news! `phpdocumentor/reflection` 5.3 was just released, and I have reverted the `${GITHUB_TOKEN}` changes from the workflows.

bshaffer commented:
The package test is failing in `ErrorReporting` with the following:

```
PHP Fatal error:  Declaration of Google\Cloud\Logging\PsrLogger::emergency($message, array $context = []) must be compatible with Psr\Log\LoggerInterface::emergency(Stringable|string $message, array $context = []): void in /home/runner/work/google-cloud-php/google-cloud-php/Logging/src/PsrLogger.php on line 172
```

bshaffer commented on Logging/composer.json:
```diff
@@ -12,23 +12,21 @@
     },
     "provide": {
-        "psr/log-implementation": "1.0|2.0"
+        "psr/log-implementation": "^1.0|^2.0|^3.0"
```
This is not accurate because `PsrLogger` is not compatible with `psr/log: 3.0` (yet)
```suggestion
        "psr/log-implementation": "^1.0|^2.0"
```

jeromegamez replied:
I just reverted and pushed the changes

bshaffer commented on Logging/composer.json:
```diff
@@ -12,23 +12,21 @@
-        "psr/log": "For using the PSR logger. Currently supports versions 1 and 2."
-    },
-    "conflict": {
-        "psr/log": ">=3"
```
This is still the case, until we update `PsrLogger` to be compatible with the interface from `psr/log: 3`, which we can't do until we drop support for PHP 7.0 and below. So this should be added back.

jeromegamez replied:
This one was just about the duplicate `"conflict"` section 😅

bshaffer reviewed (commented):
We need to remove the compatibility with `psr/log 3.0` and add back the conflicts. Fortunately, this will not affect our compatibility with `monolog/monolog`, because it has compatibility with `psr/log 2.0` as well (see [here](https://github.com/Seldaek/monolog/blob/3.2.0/composer.json#L17))

jeromegamez commented:
I have a solution for making it compatible with all three `psr/log` versions 90% working locally because I had a feeling people would demand it. I will propose this in a separate PR based on the changes in this one 🤞. I hope the latest push will finally do the trick ☺️ 

I have another question - at the moment, `run-package-tests.sh` doesn't output the installed dependencies so that we won't see which Monolog version was actually installed (same with Guzzle). What would you think about adding `composer show` to the output or removing the `-q` parameter on installation? Or would that blow the output up too much?

jeromegamez commented:
It went through! 🤯 🥳

jeromegamez commented:
@bshaffer With the changed testing process, would you like me to try reverting the changes related to components other than Core and Logging? That would lead to new PRs though, I don't know how exhausted you guys are of me already 😬.

bshaffer commented on dev/src/DocGenerator/DocGenerator.php:
```diff
@@ -82,7 +82,9 @@ public function generate($basePath, $pretty)
-        $reflection = new ReflectionHandlerV4();
+
+
+        $reflection = ReflectionHandlerFactory::create();
```
This part needs to be reverted!

jeromegamez replied:
Done ✅

bshaffer commented on Spanner/src/Session/Session.php:
```diff
@@ -185,7 +185,7 @@ public function expiration()
     {
         return [
-            'connection' => get_class($this->connection),
+            'connection' => is_object($this->connection) ? get_class($this->connection) : null,
```
Could you tell me again what this does? We don't mind keeping it if it's fixing an error, but since all tests were passing without it, I am assuming this was coming from somewhere else.

Now that we've made a bunch of changes, do you think we can revert this?

jeromegamez replied:
I believe this was from when I executed all tests from the repo's root, I just tested this again, and it works without the change. I dropped the commit and will look over the other changes as well.

bshaffer commented on .gitignore:
```diff
@@ -2,10 +2,12 @@
 build/
 composer.phar
 composer.lock
+composer-local.*
```
nice, this is better :)

updated my PR!

bshaffer commented on .github/run-package-tests.sh:
```diff
@@ -13,8 +13,29 @@
+        if grep -q "\"google/cloud-$1\":" ${DIR}/composer.json; then
+            composer config repositories.$1 "{\"type\": \"path\", \"url\": \"../$2\", \"options\":{\"versions\":{\"google/cloud-$1\":\"1.100\"}}}" -d ${DIR}
+        fi
+    done
```
We'll be sure to merge #5542 before merging this PR, so that these changes won't need to be a part of this PR.

bshaffer reviewed (commented):
This is looking really good! Just a few comments on some of the remaining changes, and once we get @dwsupplee to approve, and we merge #5542, we'll be all set!

jeromegamez commented:
I removed almost all commits that are not strictly related to the feature and renamed/reordered the others to read better and to have a fix/chore/feat order. 💪 

This PR has had quite a journey! ☺️

jeromegamez commented on Logging/composer.json:
```diff
@@ -48,8 +49,5 @@
             "Google\\Cloud\\Logging\\Tests\\": "tests"
         }
-    },
-    "conflict": {
-        "psr/log": ">=3"
```
The diff doesn't show it, but the `config` block was duplicated.

bshaffer replied:
I see that now! Thanks for pointing that out

jeromegamez commented:
@bshaffer I've seen that you merged main into this branch after #5442 but I rebased anyway to reduce the number of commits - just in case you didn't want to squash-merge later.

dwsupplee commented on Logging/src/LogMessageProcessor.php:
```diff
@@ -0,0 +1,33 @@
+/**
+ * Interface representing a Message Processor used by the {@see PsrLogger}
+ */
+interface LogMessageProcessor
```
We've trended towards suffixing interfaces with `Interface`, would you be open to doing the same here?

jeromegamez replied:
Sure!

Really just out of curiosity, I surely don't want to start a flame war 😅: what made you trend towards it? (The old "`interface x` and `y implements x` already tells that it is an interface" argument)

dwsupplee replied:
I've found it nice while browsing the file tree to be able to quickly identify things like traits/interfaces vs. classes.

dwsupplee commented on Core/src/Exception/ServiceException.php:
```diff
@@ -57,13 +57,13 @@ class ServiceException extends GoogleException
      */
     public function __construct(
-        $message,
+        $message = null,
```
What necessitates this change?

jeromegamez replied:
I tried to explain this in https://github.com/googleapis/google-cloud-php/commit/353a0b4523f832a1436da62c3a626908f1516df6

I don't know if the commit is still available when I force-push the suggested changes, so I'll copy the commit message here:

Allow `ServiceException` creation with null as the message

Some tests explicitly set the message to `null` which breaks the
creation of parent exceptions with newer PHP versions.

Setting a default `$message = ''` does not prevent the problem, so
we declare it with `null` being allowed, and check and replace it in the
`ServiceException` constructor.

The change is made in the `ServiceException` instead of the tests
calling it incorrectly to reduce the scope of updated files in the
context of the PR that handles the problem.

dwsupplee reviewed (changes requested):
My comments are mostly housekeeping related, structurally this looks great! Thanks @jeromegamez! 👍

jeromegamez commented on Core/src/Logger/AppEngineFlexFormatterV3.php:
```diff
@@ -0,0 +1,53 @@
+<?php
+/*
+ * Copyright 2019 Google LLC
```
🤦🏼‍♂️

dwsupplee commented on Logging/composer.json:
```diff
@@ -12,8 +12,9 @@
-        "phpdocumentor/reflection": "^3.0||^4.0",
+        "phpdocumentor/reflection": "^3.0||^4.0||^5.3",
         "erusev/parsedown": "^1.6",
+        "fig/log-test": "^1.0",
```
Is this an artifact that can be removed? I may have missed it, but couldn't find it being used.

jeromegamez replied:
Oh, yes! Must have been an orphan from an earlier, then reverted change.

Unfortunately I'm currently on the road, so I'll only be able to apply the removal in a few hours...

dwsupplee replied:
Never mind this comment, I hadn't realized this would be needed for https://github.com/googleapis/google-cloud-php/blob/077810260b58f5de8a3bbdfd999a5e9a48f71a7f/Logging/tests/Unit/PsrLoggerCompatibilityTest.php to pass when using psr/log:2 since the distributed package doesn't include the test classes like psr/log:1 does. Sorry for the churn, and thanks again for all the work here!

bshaffer commented:
We missed `Debugger`: https://github.com/googleapis/google-cloud-php/pull/5711

bshaffer commented:
We are going to wait until after the new year to release this, as our team is currently in a release freeze for the holidays!

jeromegamez commented:
Super glad to see it released now! I enjoyed the process! We'll meet again at latest when Monolog 4 is released 😄
